File: common/src/main/java/demo/mesharena/common/TracingHeaders.java
package demo.mesharena.common;

import io.opentracing.SpanContext;
import io.opentracing.Tracer;
import io.opentracing.propagation.Format;
import io.opentracing.propagation.TextMap;
import io.vertx.core.MultiMap;
import org.jetbrains.annotations.NotNull;

import java.util.Iterator;
import java.util.Map.Entry;

/**
 * @author Pavol Loffay
 */
public class TracingHeaders {
  public static void inject(Tracer tracer, SpanContext spanContext, MultiMap headers) {
    tracer.inject(spanContext, Format.Builtin.HTTP_HEADERS, new TextMap() {
      @NotNull
      @Override
      public Iterator<Entry<String, String>> iterator() {
        throw new UnsupportedOperationException();
      }
      @Override
      public void put(String key, String value) {
        headers.add(key, value);
      }
    });
  }

  public static SpanContext extract(Tracer tracer, MultiMap headers) {
    return tracer.extract(Format.Builtin.HTTP_HEADERS, new TextMap() {
      @NotNull
      @Override
      public Iterator<Entry<String, String>> iterator() {
        return headers.iterator();
      }
      @Override
      public void put(String key, String value) {
        throw new UnsupportedOperationException();
      }
    });
  }
}

File: common/src/main/kotlin/demo/mesharena/common/TracingHeaders.kt
package demo.mesharena.common

import io.opentracing.SpanContext
import io.opentracing.Tracer
import io.opentracing.propagation.Format
import io.opentracing.propagation.TextMap
import io.vertx.core.MultiMap

/**
 * @author Pavol Loffay
 */
object TracingHeaders {
    fun inject(tracer: Tracer, spanContext: SpanContext, headers: MultiMap) {
        tracer.inject(spanContext, Format.Builtin.HTTP_HEADERS, object : TextMap {
            override fun iterator(): MutableIterator<Map.Entry<String, String>> {
                throw UnsupportedOperationException()
            }

            override fun put(key: String, value: String) {
                headers.add(key, value)
            }
        })
    }

    fun extract(tracer: Tracer, headers: MultiMap): SpanContext {
        return tracer.extract(Format.Builtin.HTTP_HEADERS, object : TextMap {
            override fun iterator(): MutableIterator<Map.Entry<String, String>> {
                return headers.iterator()
            }

            override fun put(key: String, value: String) {
                throw UnsupportedOperationException()
            }
        })
    }
}
